File: scripts/generate-deca-questions.mjs
import { invokeLLM } from "../server/_core/llm.js";
import * as fs from "fs";
import * as path from "path";

// DECA Clusters and their Performance Indicators
const CLUSTERS = {
  marketing: {
    name: "Marketing",
    pis: [
      "Market Segmentation and Targeting",
      "Marketing Mix (Product, Price, Place, Promotion)",
      "Consumer Behavior and Market Research",
      "Channel Management and Distribution",
      "Pricing Strategies",
      "Promotional Campaigns and Advertising",
      "Brand Management and Positioning",
      "Product Lifecycle Management",
      "Selling Techniques and Customer Service",
      "Marketing Ethics and Regulations",
    ],
    questionsPerPI: 600,
  },
  finance: {
    name: "Finance",
    pis: [
      "Financial Statement Analysis",
      "Ratio Analysis",
      "Accounting Principles",
      "Cost Analysis",
      "Investment Analysis",
      "Time Value of Money",
      "Risk Management and Insurance",
      "Credit and Debt Management",
      "Tax Implications",
      "Financial Forecasting",
    ],
    questionsPerPI: 600,
  },
  businessAdminCore: {
    name: "Business Administration Core",
    pis: [
      "Business Law and Regulations",
      "Communication Skills",
      "Customer Relations",
      "Economics",
      "Emotional Intelligence",
      "Entrepreneurship",
      "Finance Fundamentals",
      "Human Resources Management",
      "Information Management",
      "Marketing Basics",
      "Operations Management",
      "Professional Development",
      "Strategic Management",
    ],
    questionsPerPI: 460,
  },
  hospitality: {
    name: "Hospitality & Tourism",
    pis: [
      "Revenue Management",
      "Customer Service Excellence",
      "Hospitality Operations",
      "Food and Beverage Management",
      "Housekeeping and Maintenance",
      "Front Office Operations",
      "Sales and Marketing for Hospitality",
      "Human Resources in Hospitality",
      "Financial Management",
      "Quality Assurance",
      "Risk Management",
      "Sustainability",
    ],
    questionsPerPI: 500,
  },
  management: {
    name: "Management",
    pis: [
      "Human Resources Management",
      "Organizational Behavior",
      "Leadership and Motivation",
      "Team Dynamics",
      "Performance Management",
      "Compensation and Benefits",
      "Recruitment and Selection",
      "Training and Development",
      "Labor Relations",
      "Strategic Management",
      "Operations Management",
      "Decision-Making",
    ],
    questionsPerPI: 500,
  },
  entrepreneurship: {
    name: "Entrepreneurship",
    pis: [
      "Business Planning",
      "Market Research",
      "Financial Projections",
      "Product Development",
      "Marketing Strategies",
      "Operations Planning",
      "Risk Assessment",
      "Legal and Regulatory Requirements",
      "Competitive Analysis",
      "Growth Strategies",
      "Innovation",
    ],
    questionsPerPI: 545,
  },
  pfl: {
    name: "Personal Financial Literacy",
    pis: [
      "Income and Earning Potential",
      "Budgeting and Spending",
      "Saving and Investment",
      "Credit and Debt",
      "Insurance and Risk",
      "Retirement Planning",
      "Tax Planning",
      "Financial Goals",
      "Consumer Protection",
      "Financial Decision-Making",
    ],
    questionsPerPI: 400,
  },
};

const DIFFICULTIES = ["Beginner", "Intermediate", "Advanced"];
const QUESTION_TYPES = [
  "Concept Definition",
  "Scenario-Based Decision",
  "Calculation Analysis",
  "Comparison Exclusion",
];

async function generateQuestionsForPI(
  cluster,
  clusterName,
  pi,
  count,
  difficulty
) {
  const prompt = `You are a DECA exam question generator. Generate exactly 5 high-quality multiple-choice questions for the ${clusterName} cluster, specifically testing the Performance Indicator: "${pi}".

Requirements:
1. Each question must be authentic to DECA standards and use professional business terminology
2. Questions should test real-world business scenarios and decision-making
3. Difficulty level: ${difficulty}
4. Format each question as JSON with this structure:
{
  "question": "The question text here?",
  "options": ["Option A", "Option B", "Option C", "Option D"],
  "correctAnswerIndex": 0,
  "explanation": "Detailed explanation of why this is correct..."
}

5. Vary question types: mix concept definitions, scenario-based decisions, calculations, and comparisons
6. Use DECA-appropriate examples and business contexts
7. Ensure each question has one clear correct answer and plausible distractors
8. For calculations, include realistic numbers and show the methodology in the explanation

Generate 5 questions now. Return ONLY a valid JSON array with 5 question objects, no additional text.`;

  try {
    const response = await invokeLLM({
      messages: [
        {
          role: "system",
          content:
            "You are a DECA exam expert creating authentic practice questions. Return only valid JSON arrays.",
        },
        {
          role: "user",
          content: prompt,
        },
      ],
    });

    const content = response.choices[0].message.content;
    const jsonMatch = content.match(/\[[\s\S]*\]/);

    if (!jsonMatch) {
      console.error(`Failed to extract JSON for ${clusterName} - ${pi}`);
      return [];
    }

    const questions = JSON.parse(jsonMatch[0]);
    return questions.map((q) => ({
      ...q,
      cluster,
      clusterName,
      performanceIndicator: pi,
      difficulty,
    }));
  } catch (error) {
    console.error(`Error generating questions for ${pi}:`, error.message);
    return [];
  }
}

async function generateAllQuestions() {
  const allQuestions = [];
  let totalGenerated = 0;

  console.log("Starting DECA question generation...\n");

  for (const [clusterKey, clusterData] of Object.entries(CLUSTERS)) {
    console.log(`\n📚 Processing ${clusterData.name} cluster...`);

    for (const pi of clusterData.pis) {
      console.log(`  📖 Generating questions for: ${pi}`);

      // Generate questions for each difficulty level
      for (const difficulty of DIFFICULTIES) {
        const questionsPerDifficulty = Math.ceil(
          clusterData.questionsPerPI / DIFFICULTIES.length
        );

        const questions = await generateQuestionsForPI(
          clusterKey,
          clusterData.name,
          pi,
          questionsPerDifficulty,
          difficulty
        );

        allQuestions.push(...questions);
        totalGenerated += questions.length;

        console.log(
          `    ✓ Generated ${questions.length} ${difficulty} questions`
        );

        // Add delay to avoid rate limiting
        await new Promise((resolve) => setTimeout(resolve, 500));
      }
    }
  }

  console.log(`\n✅ Total questions generated: ${totalGenerated}`);
  return allQuestions;
}

async function saveQuestionsToFile(questions) {
  const outputDir = path.join(process.cwd(), "data");
  if (!fs.existsSync(outputDir)) {
    fs.mkdirSync(outputDir, { recursive: true });
  }

  const outputFile = path.join(outputDir, "deca-questions.json");
  fs.writeFileSync(outputFile, JSON.stringify(questions, null, 2));

  console.log(`\n💾 Questions saved to: ${outputFile}`);
  console.log(`   Total questions: ${questions.length}`);

  // Generate summary statistics
  const summary = {
    totalQuestions: questions.length,
    byCluster: {},
    byDifficulty: {},
  };

  for (const q of questions) {
    summary.byCluster[q.clusterName] =
      (summary.byCluster[q.clusterName] || 0) + 1;
    summary.byDifficulty[q.difficulty] =
      (summary.byDifficulty[q.difficulty] || 0) + 1;
  }

  console.log("\n📊 Summary Statistics:");
  console.log("By Cluster:", summary.byCluster);
  console.log("By Difficulty:", summary.byDifficulty);
}

async function main() {
  try {
    const questions = await generateAllQuestions();
    await saveQuestionsToFile(questions);
    console.log("\n🎉 Question generation complete!");
  } catch (error) {
    console.error("Fatal error:", error);
    process.exit(1);
  }
}

main();

File: scripts/generate_deca_questions.py
#!/usr/bin/env python3
"""
DECA Question Generation Script
Generates 40,000+ authentic DECA-aligned practice questions
"""

import json
import os
import sys
import time
import requests
from pathlib import Path

# Get API credentials
FORGE_API_URL = os.environ.get("BUILT_IN_FORGE_API_URL")
FORGE_API_KEY = os.environ.get("BUILT_IN_FORGE_API_KEY")

if not FORGE_API_URL or not FORGE_API_KEY:
    print("❌ Missing BUILT_IN_FORGE_API_URL or BUILT_IN_FORGE_API_KEY environment variables")
    sys.exit(1)

# DECA Structure
DECA_STRUCTURE = {
    "marketing": {
        "name": "Marketing",
        "totalQuestions": 6000,
        "performanceIndicators": [
            {
                "name": "Market Segmentation and Targeting",
                "description": "Dividing markets into distinct groups and selecting target segments",
                "questionsNeeded": 600,
            },
            {
                "name": "Marketing Mix - Product",
                "description": "Product development, features, quality, and lifecycle management",
                "questionsNeeded": 500,
            },
            {
                "name": "Marketing Mix - Price",
                "description": "Pricing strategies, elasticity, discounting, and value perception",
                "questionsNeeded": 500,
            },
            {
                "name": "Marketing Mix - Place",
                "description": "Distribution channels, logistics, and supply chain management",
                "questionsNeeded": 500,
            },
            {
                "name": "Marketing Mix - Promotion",
                "description": "Advertising, sales promotion, public relations, and personal selling",
                "questionsNeeded": 500,
            },
            {
                "name": "Consumer Behavior and Market Research",
                "description": "Understanding customer needs, preferences, and decision-making",
                "questionsNeeded": 500,
            },
            {
                "name": "Brand Management and Positioning",
                "description": "Building brand identity, equity, and competitive positioning",
                "questionsNeeded": 450,
            },
            {
                "name": "Selling Techniques and Customer Service",
                "description": "Sales process, customer retention, and service excellence",
                "questionsNeeded": 450,
            },
            {
                "name": "Marketing Ethics and Regulations",
                "description": "Ethical practices, consumer protection, and legal compliance",
                "questionsNeeded": 400,
            },
            {
                "name": "Marketing Analysis and Planning",
                "description": "SWOT analysis, market forecasting, and strategic planning",
                "questionsNeeded": 400,
            },
        ],
    },
    "finance": {
        "name": "Finance",
        "totalQuestions": 6000,
        "performanceIndicators": [
            {
                "name": "Financial Statement Analysis",
                "description": "Analyzing income statements, balance sheets, and cash flow statements",
                "questionsNeeded": 700,
            },
            {
                "name": "Ratio Analysis and Interpretation",
                "description": "Liquidity, profitability, efficiency, and leverage ratios",
                "questionsNeeded": 700,
            },
            {
                "name": "Accounting Principles and Practices",
                "description": "GAAP, journal entries, ledgers, and financial reporting",
                "questionsNeeded": 600,
            },
            {
                "name": "Cost Analysis and Management",
                "description": "Fixed vs variable costs, break-even analysis, and cost control",
                "questionsNeeded": 600,
            },
            {
                "name": "Investment Analysis and Valuation",
                "description": "Stock and bond analysis, NPV, IRR, and investment decisions",
                "questionsNeeded": 600,
            },
            {
                "name": "Time Value of Money",
                "description": "Present value, future value, annuities, and compound interest",
                "questionsNeeded": 500,
            },
            {
                "name": "Risk Management and Insurance",
                "description": "Risk assessment, mitigation strategies, and insurance types",
                "questionsNeeded": 450,
            },
            {
                "name": "Credit and Debt Management",
                "description": "Loans, bonds, credit terms, and debt restructuring",
                "questionsNeeded": 450,
            },
            {
                "name": "Tax Planning and Implications",
                "description": "Tax types, deductions, credits, and tax-efficient strategies",
                "questionsNeeded": 400,
            },
            {
                "name": "Financial Forecasting and Budgeting",
                "description": "Projections, variance analysis, and budget management",
                "questionsNeeded": 400,
            },
        ],
    },
}

DIFFICULTIES = ["Beginner", "Intermediate", "Advanced"]


def invoke_llm(messages):
    """Call the LLM API to generate questions"""
    payload = {
        "model": "gpt-4o-mini",
        "messages": messages,
        "temperature": 0.7,
        "max_tokens": 4000,
    }

    headers = {
        "Content-Type": "application/json",
        "Authorization": f"Bearer {FORGE_API_KEY}",
    }

    response = requests.post(
        f"{FORGE_API_URL}/v1/chat/completions",
        json=payload,
        headers=headers,
        timeout=60,
    )

    if response.status_code != 200:
        raise Exception(
            f"LLM API error: {response.status_code} {response.text}"
        )

    return response.json()


def generate_questions_for_pi(cluster, pi, difficulty, count):
    """Generate questions for a specific Performance Indicator"""
    prompt = f"""You are an expert DECA exam question generator with deep knowledge of DECA's official exam standards.

Generate exactly {count} authentic, high-quality multiple-choice questions for the {cluster['name']} cluster.

Performance Indicator: "{pi['name']}"
Description: {pi['description']}
Difficulty Level: {difficulty}

CRITICAL REQUIREMENTS:
1. Use DECA's official terminology and business language
2. Test real-world business scenarios and decision-making
3. Each question must have ONE clear correct answer
4. Include plausible distractors that test common misconceptions
5. Questions must vary in type: definitions, calculations, scenarios, comparisons
6. For calculations: use realistic business numbers
7. Ensure questions are at the specified difficulty level
8. Each question must genuinely test the specified Performance Indicator

Return ONLY a valid JSON array with exactly {count} question objects. Each object must have this exact structure:
{{
  "question": "The question text?",
  "options": ["Option A", "Option B", "Option C", "Option D"],
  "correctAnswerIndex": 0,
  "explanation": "Detailed explanation..."
}}

Generate {count} questions now. Return ONLY the JSON array, no other text."""

    try:
        response = invoke_llm(
            [
                {
                    "role": "system",
                    "content": "You are a DECA exam expert. Generate only valid JSON arrays of questions.",
                },
                {
                    "role": "user",
                    "content": prompt,
                },
            ]
        )

        content = response["choices"][0]["message"]["content"]
        
        # Extract JSON from response
        import re
        json_match = re.search(r'\[[\s\S]*\]', content)
        
        if not json_match:
            print(f"❌ Failed to extract JSON for {cluster['name']} - {pi['name']} ({difficulty})")
            return []

        questions = json.loads(json_match.group())
        return [
            {
                **q,
                "cluster": cluster["name"],
                "performanceIndicator": pi["name"],
                "difficulty": difficulty,
            }
            for q in questions
        ]

    except Exception as e:
        print(f"❌ Error generating questions for {pi['name']}: {str(e)}")
        return []


def generate_all_questions():
    """Generate all DECA questions"""
    all_questions = []
    total_generated = 0
    start_time = time.time()

    print("🚀 Starting DECA question generation...\n")

    for cluster_key, cluster in DECA_STRUCTURE.items():
        print(f"\n📚 {cluster['name']} Cluster (Target: {cluster['totalQuestions']} questions)")

        for pi in cluster["performanceIndicators"]:
            print(f"  📖 {pi['name']}")

            questions_per_difficulty = (pi["questionsNeeded"] + 2) // 3

            for difficulty in DIFFICULTIES:
                questions = generate_questions_for_pi(
                    cluster, pi, difficulty, questions_per_difficulty
                )

                all_questions.extend(questions)
                total_generated += len(questions)

                print(
                    f"    ✓ {difficulty}: {len(questions)} questions (Total: {total_generated})"
                )

                # Rate limiting
                time.sleep(1)

    elapsed_time = (time.time() - start_time) / 60
    print(f"\n✅ Generation complete in {elapsed_time:.2f} minutes")
    print(f"📊 Total questions generated: {total_generated}")

    return all_questions


def save_questions(questions):
    """Save questions to file"""
    output_dir = Path("data")
    output_dir.mkdir(exist_ok=True)

    output_file = output_dir / "deca-questions.json"
    with open(output_file, "w") as f:
        json.dump(questions, f, indent=2)

    print(f"\n💾 Questions saved to: {output_file}")

    # Generate statistics
    stats = {
        "totalQuestions": len(questions),
        "byCluster": {},
        "byDifficulty": {},
    }

    for q in questions:
        cluster = q["cluster"]
        difficulty = q["difficulty"]
        stats["byCluster"][cluster] = stats["byCluster"].get(cluster, 0) + 1
        stats["byDifficulty"][difficulty] = stats["byDifficulty"].get(difficulty, 0) + 1

    print("\n📊 Statistics:")
    print("By Cluster:", stats["byCluster"])
    print("By Difficulty:", stats["byDifficulty"])


def main():
    try:
        questions = generate_all_questions()
        save_questions(questions)
        print("\n🎉 Question generation complete!")
    except Exception as e:
        print(f"❌ Fatal error: {str(e)}")
        sys.exit(1)


if __name__ == "__main__":
    main()
